src/replication.c

```diff
@@ -1626,53 +1669,60 @@ timeout_wait_for_command(spec_t *spec, rcommon_mgmt_cmd_t *rcomm_mgmt, int wait_
 }
 
 /*
- * This API will create snapshot with given name on the spec.
+ * This API will create snapshot or resize volume based on the opcode.
  * It will wait for io_wait_time seconds to complete ongoing IOs.
  * Overall, this API will wait for wait_time seconds to get response
  * for snapshot command (this includes io_wait_time).
  * In case of any failures, snapshot destroy command will be sent to all replicas.
  */
-int istgt_lu_create_snapshot(spec_t *spec, char *snapname, int io_wait_time, int wait_time)
+int istgt_execute_volume_operation(spec_t *spec, zvol_op_code_t opcode,
+			char *snapname, uint64_t size, int io_wait_time, int wait_time)
 {
 	bool r;
-	int ret = false, sent = 0, success;
 	replica_t *replica;
+	int sent = 0, success, ret;
+	uint8_t rf, cf;
 	struct timespec last;
 	rcommon_mgmt_cmd_t *rcomm_mgmt;
 	uint64_t io_seq;
+	size_t data_len = 0;
+	void *data = NULL;
+	zvol_op_resize_data_t *resize_cmd;
+
 
 	clock_gettime(CLOCK_MONOTONIC_COARSE, &last);
 	MTX_LOCK(&spec->rq_mtx);
 
-	/* Wait for any ongoing snapshot commands */
+	/* Wait for any ongoing commands which colud block IOs */
 	while (spec->quiesce == 1) {
 		MTX_UNLOCK(&spec->rq_mtx);
 		sleep(1);
 		MTX_LOCK(&spec->rq_mtx);
 	}
 
-	if (can_take_snapshot(spec) == false) {
+	if (can_send_opcode(spec, opcode) == false) {
 		MTX_UNLOCK(&spec->rq_mtx);
-		REPLICA_ERRLOG("volume is not healthy to take snapshots..\n");
+		REPLICA_ERRLOG("Volume is not healthy to process OPCODE: %d", opcode);
 		return false;
 	}
 
-	r = pause_and_timed_wait_for_ongoing_ios(spec, io_wait_time);
+	r = pause_and_timed_wait_for_ongoing_ios(spec, opcode, io_wait_time);
 	if (r == false) {
 		MTX_UNLOCK(&spec->rq_mtx);
 		REPLICA_ERRLOG("pausing failed..\n");
 		return false;
 	}
 
 	io_seq = ++spec->io_seq;
-	uint8_t cf = spec->consistency_factor;
-	uint8_t rf = spec->replication_factor;
+	cf = spec->consistency_factor;
+	rf = spec->replication_factor;
 
 	rcommon_mgmt_cmd_t *rmgmt = allocate_rcommon_mgmt_cmd(0);
 	replica_t *hr = spec->rebuild_info.healthy_replica;
-	if (hr) {
+	if (hr && opcode == ZVOL_OPCODE_SNAP_CREATE) {
+		BUILD_VOL_CMD_DATA(spec)
```

Comment:
have this kind of macro defined as a function call.
 https://github.com/torvalds/linux/blob/ecb095bff5d4b8711a81968625b3b4a235d3e477/include/asm-generic/pgtable.h#L411